media/sse/src/main/java/org/glassfish/jersey/media/sse/internal/JerseyEventSink.java

```diff
@@ -39,55 +45,109 @@
+
+    private static final Map<Class<?>, MediaType> IMPLICIT_MEDIA_TYPE;
+
+    static {
+        Map<Class<?>, MediaType> map = new HashMap<>(8);
+
+        // Already wrapped
+        map.put(OutboundSseEvent.class, MediaType.SERVER_SENT_EVENTS_TYPE);
+
+        // No wrapping for raw types
+        map.put(String.class, MediaType.TEXT_PLAIN_TYPE);
+        map.put(Character.class, MediaType.TEXT_PLAIN_TYPE);
+        map.put(Boolean.class, MediaType.TEXT_PLAIN_TYPE);
+        map.put(Byte.class, MediaType.TEXT_PLAIN_TYPE);
+        map.put(Short.class, MediaType.TEXT_PLAIN_TYPE);
+        map.put(Integer.class, MediaType.TEXT_PLAIN_TYPE);
+        map.put(Long.class, MediaType.TEXT_PLAIN_TYPE);
+        map.put(Double.class, MediaType.TEXT_PLAIN_TYPE);
+
+        IMPLICIT_MEDIA_TYPE = Collections.unmodifiableMap(map);
+    }
 
     JerseyEventSink(Provider<AsyncContext> asyncContextProvider) {
         super(SSE_EVENT_DELIMITER, asyncContextProvider);
     }
 
     @Override
     public void onSubscribe(final Flow.Subscription subscription) {
-        checkClosed();
         if (subscription == null) {
             throw new NullPointerException(LocalizationMessages.PARAM_NULL("subscription"));
         }
+        if (subscribed.getAndSet(true)) {
+            subscription.cancel();
+            return;
+        }
+
         this.subscription = subscription;
-        subscription.request(Long.MAX_VALUE);
+        if (isClosed()) {
+            subscription.cancel();
+        } else {
+            subscription.request(Long.MAX_VALUE);
+        }
     }
 
 
     @Override
-    public void onNext(final OutboundSseEvent item) {
-        checkClosed();
+    public void onNext(final Object item) {
         if (item == null) {
             throw new NullPointerException(LocalizationMessages.PARAM_NULL("outboundSseEvent"));
         }
         try {
-            write(item);
-        } catch (final IOException e) {
-            onError(e);
+            checkClosed();
+            MediaType implicitType = resolveMediaType(item);
+            if (MediaType.SERVER_SENT_EVENTS_TYPE == implicitType) {
+                // already wrapped
+                write((OutboundSseEvent) item);
+            } else {
+                // implicit wrapping
+                // TODO: Jersey annotation for explicit media type
+                write(new OutboundEvent.Builder()
+                        .mediaType(implicitType)
+                        .data(item)
+                        .build());
+            }
+        } catch (final Throwable e) {
+            LOGGER.log(Level.INFO, LocalizationMessages.EVENT_SOURCE_DEFAULT_ONERROR(), e);
         }
     }
 
     @Override
     public void onError(final Throwable throwable) {
-        checkClosed();
         if (throwable == null) {
             throw new NullPointerException(LocalizationMessages.PARAM_NULL("throwable"));
         }
-        subscription.cancel();
+        try {
+            super.close();
```

Comment:
how about calling this.close()? Same as onComplete? Would you still need the onClose()?

Reply:
Intention was differentiate client side close(disconnect) where `cancel` is expected, from upstream close where `cancel` is not desirable as upstream is cause of closing. Change should avoid similar issues as oracle/helidon/issues/806 in the future.